### ISSUE TITLE
Combine "monotonic" and "primary key" analyses

### ISSUE THREAD

**Issue by frankmcsherry** [MaterializeInc/materialize#4298]
We recently started using information about "monotonicity" or "append-only" -ness of an input stream. These collections only ever add records, and this results in some streamlined implementations for certain aggregations (e.g. top k, min, max). We could extend this to more optimizations if we include a peek at the existence of primary keys for these inputs.

Specifically, if a stream is both append-only and has primary keys, we can think of it as "write once" for each of its keys: once we see a `(key, val)` pair, we should never see that `key` again.

This leads to at least one appealing optimization, when we join an append-only stream on a primary key. Let's imagine we have input collections `A` and `B`, and `A` is append-only and has a primary key that we are using in the join. Generally, we need to maintain all of `A` and all of `B`, indefinitely. However, in this specific case, for any matching records `(key, a, b)` we can retract `b` from the arrangement maintaining `B`, as it will never match another record in `A`. Effectively, the arrangement for `B` only needs to maintain *unmatched* records for `B`.

For an example of where this applies, consider the problem of matching records corresponding to connect begin and end events. Each connection has a primary key (say) and each begin and end event happens only once. The goal is to match the begin and end events, which are unbounded over time. However, the primary key nature of the inputs allows us to implement the join in a way that automatically retracts matches from each of the inputs, keeping the memory footprint bounded.

I believe we can implement this using existing "feedback" idioms, though planning this might fight a little with explicit arrangements in our query plans (which I think remove our ability to choose to maintain anything other than that content). Ping @wangandi on that, and whether relaxing the explicit arrangements might be permissible and what the fall-out might be.

**Comment by awang** [comment]
Are the "existing feedback idioms" you refer to in our SQL DDL (ie CREATE)?

**Comment by frankmcsherry** [comment]
No. I meant the Timely Dataflow `feedback` operator, which is used in a few places in the rendering of dataflows (e.g. `TopK` and `Reduce`) as well as in our log maintenance code.

**Comment by frankmcsherry** [comment]
Let me try and present some pseudocode that might clarify the potential optimization.

Right now we often find ourselves implementing "a join b on (a.key1 = b.key2)" or something semantically equivalent. When it comes time to render this, we usually end up performing something that looks like
```rust
let a_keyed = a.map(|row| (key1(row), row)).arrange_by_key();
let b_keyed = b.map(|row| (key2(row), row)).arrange_by_key();
let result = a_keyed.join_core(b_keyed);
```
There is usually some funny business before we get `a` and `b`, and we might have some arrangements at hand that we can use instead.

Instead, we have the option to create semantically different arrangements for each input, which may *retract* any inputs observed in the output, if the *other* input is both append-only and has a subset of the join key as a primary key.

```rust
// Captures the result of the join, to be used for retraction.
let output = Variable::new(scope, delay_ns);
let mut retracted = false;

// Retract input from `a` if `b` is write-once.
let mut a_keyed = a.map(|row| (key1(row), row));
if is_monotonic(b) && primary_key(b, key2) {
    retracted = true;
    a_keyed = 
    output
        .map(|(key, (a_part, _b_part))| (key, a_part))
        .negate()
        .concat(&a_keyed);
}
let a_keyed = a_keyed.arrange_by_key();

// Retract input from `b` if `a` is write-once.
let mut b_keyed = b.map(|row| (key2(row), row));
if is_monotonic(a) && primary_key(a, key1) {
    retracted = true;
    b_keyed = 
    output
        .map(|(key, (_a_part, b_part))| (key, b_part))
        .negate()
        .concat(&b_keyed);
}
let b_keyed = b_keyed.arrange_by_key();

let mut result = a_keyed.join_core(b_keyed);

output.bind(result);

// If input retractions occur, we need to avoid 
// retracting the output.
if retracted {
    result = output.concat(&result);
}
```
